src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,173 @@
+            head = newNode;
+        } else {
+            Node<T> current = head;
+            for (int i = 0; i < index - 1; i++) {
+                current = current.next;
+            }
+            newNode.next = current.next;
+            current.next = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T item : list) {
+            add(item);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index out of bounds");
+        }
+
+        Node<T> current = head;
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
+        return current.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index out of bounds");
+        }
+
+        Node<T> current = head;
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
+
+        T oldValue = current.value;
+        current.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index out of bounds");
+        }
+
+        Node<T> current = head;
+        Node<T> prev = null;
+
+        if (index == 0) {
+            T value = head.value;
+            head = head.next;
+            size--;
+            return value;
+        }
+
+        for (int i = 0; i < index; i++) {
+            prev = current;
+            current = current.next;
+        }
+
+        prev.next = current.next;
+        size--;
+        return current.value;
     }
 
     @Override
     public boolean remove(T object) {
+        if (head == null) {
+            return false;
+        }
+
+        Node<T> current = head;
+        while (current != null) {
+            if (Objects.equals(current.value, object)) {
+                if (current == head) {
+                    head = head.next;
+                    if (head != null) {
+                        head.prev = null;
+                    } else {
+                        tail = null;
+                    }
+                } else if (current == tail) {
+                    tail = current.prev;
+                    if (tail != null) {
+                        tail.next = null;
+                    }
+                } else {
+                    if (current.prev != null) {
+                        current.prev.next = current.next;
+                    }
+                    if (current.next != null) {
+                        current.next.prev = current.prev;
+                    }
+                }
+                size--;
+                return true;
+            }
+            current = current.next;
+        }
```

Comment:
Don't iterate your nodes from the head to the tail without considering the position of the index. You can iterate only the first or second half of the list, depending on the index value, to boost performance.